### PR TITLE
WIP: bare metal

### PR DESCRIPTION
**Pull request by justinsb**
Very WIP at the current time - mostly just splitting up the branches

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2767)

### CODE REVIEW

File: pkg/apis/kops/instancegroup.go [hunk 1]
```diff
@@ -103,6 +105,15 @@ type InstanceGroupSpec struct {
 	Taints []string `json:"taints,omitempty"`
 }
 
+type MachineSpec struct {
```
Comment (chrislovecnm) [inline pkg/apis/kops/instancegroup.go]: Thought you did not like the word Spec inside of a Spec.

File: pkg/apis/kops/instancegroup.go [hunk 2]
```diff
@@ -144,3 +155,59 @@ func (g *InstanceGroup) IsMaster() bool {
 		return false
 	}
 }
+
+func (g *InstanceGroup) Validate() error {
```
Comment (chrislovecnm) [inline pkg/apis/kops/instancegroup.go]: Should the validate code go here or inside the validate pkg?

File: pkg/apis/kops/instancegroup.go (hunk 1 above)
Comment (justinsb) [inline pkg/apis/kops/instancegroup.go]: The struct doesn't much matter, it is the field name (and strictly only the json form of it)

File: pkg/apis/kops/instancegroup.go (hunk 2 above)
Comment (justinsb) [inline pkg/apis/kops/instancegroup.go]: You're right, probably in the validate pkg

**Comment by cdrage** [comment]
Hey @justinsb is this still a work-in-progress? Is it possible to get an update to the situation.

To be honest, it's going to be *a lot* of work considering how different each bare-metal environment is as well as the effort towards `kubeadm` and `kubespray`. Do you think that this is still viable for the future?

**Comment by chrislovecnm** [comment]
@cdrage this PR probably needs to be closed. We have some initial basic support for bare metal already merged. 

Would be good if @justinsb could document it a bit ;)

**Pull request closed without merging**